internal/k8s/vcf_controller.go

```diff
@@ -0,0 +1,181 @@
+	"k8s.io/apimachinery/pkg/util/runtime"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+func (c *AviController) AddNCPBootstrapEventHandler(k8sinfo K8sinformers, stopCh <-chan struct{}, startSyncCh chan struct{}) {
+	cs := k8sinfo.Cs
+	utils.AviLog.Debugf("Creating event broadcaster for NCP Bootstrap CRD")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(utils.AviLog.Debugf)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: cs.CoreV1().Events("")})
+	NCPBootstrapHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap ADD Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+		UpdateFunc: func(old, obj interface{}) {
+			utils.AviLog.Infof("NCP Bootstrap Update Event")
+			if c.ValidBSData() && startSyncCh != nil {
+				c.UpdateAviSecret()
+				startSyncCh <- struct{}{}
+				startSyncCh = nil
+			}
+		},
+	}
+	c.dynamicInformers.NCPBootstrapInformer.Informer().AddEventHandler(NCPBootstrapHandler)
+
+	go c.dynamicInformers.NCPBootstrapInformer.Informer().Run(stopCh)
+	if !cache.WaitForCacheSync(stopCh, c.dynamicInformers.NCPBootstrapInformer.Informer().HasSynced) {
+		runtime.HandleError(fmt.Errorf("Timed out waiting for caches to sync"))
+	} else {
+		utils.AviLog.Info("Caches synced for NCP Bootstrap informer")
+	}
+}
+
+// HandleVCF checks if avi secret used by AKO is already present. If found, the it would try to connect to
+// AVI Controller. If there is any failure, we would look at Bootstrap CR used by NCP to communicate with AKO.
+// If Bootstrap CR is not found, AKO would wait for it to be created. If the authtoken from Bootstrap CR
+// can be used to connect to the AVI Controller, then avi-secret would be created with that token.
+func (c *AviController) HandleVCF(informers K8sinformers, stopCh <-chan struct{}, ctrlCh chan struct{}) {
+	cs := c.informers.ClientSet
+	aviSecret, err := cs.CoreV1().Secrets(utils.GetAKONamespace()).Get(context.TODO(), lib.AviSecret, metav1.GetOptions{})
+	if err == nil {
+		ctrlIP := os.Getenv(utils.ENV_CTRL_IPADDRESS)
+		authToken := aviSecret.Data["authToken"]
+		username := aviSecret.Data["username"]
+		var transport *http.Transport
+		_, err = clients.NewAviClient(
+			ctrlIP, string(username), session.SetAuthToken(string(authToken)),
+			session.SetRefreshAuthTokenCallbackV2(utils.GetAuthtokenFromCache),
+			session.SetNoControllerStatusCheck, session.SetTransport(transport),
+			session.SetInsecure,
+		)
+		if err == nil {
+			utils.AviLog.Infof("Successfully connected to AVI controller using existing AKO secret")
+			return
+		} else {
+			utils.AviLog.Error("AVI controller initialization failed with err: %v", err)
+		}
+	}
+
+	utils.AviLog.Infof("Got error while fetching avi-secret: %v", err)
```

Comment:
This log will be printed even if you got the secret but failed to connect to the controller.

Reply:
Thanks, I'll update the log in the else condition.